Fix emulation being performed in the wrong mode

NeatMonster opened the pull request:
After the merge of the PR referenced at the end of this message into the master
branch of Unicorn in July 2016, the emulation mode is set according to the
least significant of bit of the program counter when it is updated.

When an instruction is not implemented at Manticore's level and emulation is
used, the PC value passed to the emu_start() function will trigger a switch of
the current mode, setting it back to ARM mode instead of Thumb mode. This commit
fixes this issue by ensuring that the least significant bit of the PC is set
when Thumb mode emulation is performed.

See https://github.com/unicorn-engine/unicorn/pull/592

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/trailofbits/manticore/1240)
<!-- Reviewable:end -->

disconnect3d commented on manticore/utils/emulate.py:
```diff
@@ -248,7 +248,10 @@ def _step(self, instruction):
         try:
-            self._emu.emu_start(self._cpu.PC, self._cpu.PC + instruction.size, count=1)
+            pc = self._cpu.PC
+            if self._uc_mode == UC_MODE_THUMB:
```
Can you show a minimal example of this issue?

Also, is there any place we could move it so the general code isn't mixed with cpu-specific code? ;/

NeatMonster commented:
I've added a missing check to ensure that the target architecture is ARM.
I've also added a dump of the current instruction to the debug logger.

Here is a script you can use for testing the issue:
```
from manticore.core.cpu.arm import Armv7Cpu as Cpu
from manticore.core.memory import Memory32
from manticore.utils.emulate import UnicornEmulator

from capstone import *
from keystone import *

#####
import logging
from manticore.utils.log import all_loggers

for logger_name in all_loggers:
    logger = logging.getLogger(logger_name)
    if logger_name == "manticore.manticore" \
            or logger_name.startswith("manticore.core.smtlib"):
        logger.setLevel(logging.INFO)
    else:
        logger.setLevel(logging.DEBUG)
#####

ks = Ks(KS_ARCH_ARM, KS_MODE_THUMB)


def assemble(asm):
    ords = ks.asm(asm)[0]
    if not ords:
        raise Exception(f'bad assembly: {asm}')
    return ''.join(map(chr, ords))


def emulate_next(cpu):
    cpu.decode_instruction(cpu.PC)
    emu = UnicornEmulator(cpu)
    emu.emulate(cpu.instruction)


def setup_cpu(cpu, asm):
    code = cpu.memory.mmap(0x1000, 0x1000, 'rwx')
    data = cpu.memory.mmap(0xd000, 0x1000, 'rw')
    stack = cpu.memory.mmap(0xf000, 0x1000, 'rw')
    cpu.memory.write(code, assemble(asm))
    cpu.regfile.write('PC', code)
    cpu.regfile.write('SP', stack + 0x1000)


cpu = Cpu(Memory32())
mem = cpu.memory
reg = cpu.regfile

asm = 'add r1, r2, r3'
setup_cpu(cpu, asm)

cpu.mode = CS_MODE_THUMB
cpu._set_mode_by_val(0x1001)
cpu.PC = 0x1000

reg.write('R1', 0)
reg.write('R2', 0x1234)
reg.write('R3', 0x5678)
emulate_next(cpu)
assert reg.read('R1') == 0x68ac
```

Without the patch:
```
2018-11-01 12:53:43,218: [27153] m.c.memory:DEBUG: New memory map @1000 size:1000
2018-11-01 12:53:43,218: [27153] m.c.memory:DEBUG: New memory map @d000 size:1000
2018-11-01 12:53:43,218: [27153] m.c.memory:DEBUG: New memory map @f000 size:1000
2018-11-01 12:53:43,219: [27153] m.c.c.arm:DEBUG: swapping into THUMB mode
2018-11-01 12:53:43,220: [27153] m.u.emulate:DEBUG: 0x1000:	add.w	r1, r2, r3
2018-11-01 12:53:43,221: [27153] m.u.emulate:DEBUG: ==========
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R0   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R1   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R2   Manticore: 00001234, Unicorn 00001234
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R3   Manticore: 00005678, Unicorn 00005678
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R4   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R5   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R6   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R7   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R8   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R9   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R10  Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R11  Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,222: [27153] m.u.emulate:DEBUG: Register R12  Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,223: [27153] m.u.emulate:DEBUG: Register R13  Manticore: 00010000, Unicorn 00010000
2018-11-01 12:53:43,223: [27153] m.u.emulate:DEBUG: Register R14  Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,223: [27153] m.u.emulate:DEBUG: Register R15  Manticore: 00001000, Unicorn 00001000
2018-11-01 12:53:43,223: [27153] m.u.emulate:DEBUG: Register APSR  Manticore: 00000000, Unicorn 00000000
2018-11-01 12:53:43,223: [27153] m.u.emulate:DEBUG: >>>>>>>>>>
Traceback (most recent call last):
  File "/tmp/thumb-mode-emulation.py", line 62, in <module>
    assert reg.read('R1') == 0x68ac
AssertionError
```

With the patch:
```
2018-11-01 12:49:50,151: [26675] m.c.memory:DEBUG: New memory map @1000 size:1000
2018-11-01 12:49:50,151: [26675] m.c.memory:DEBUG: New memory map @d000 size:1000
2018-11-01 12:49:50,152: [26675] m.c.memory:DEBUG: New memory map @f000 size:1000
2018-11-01 12:49:50,152: [26675] m.c.c.arm:DEBUG: swapping into THUMB mode
2018-11-01 12:49:50,153: [26675] m.u.emulate:DEBUG: 0x1000:	add.w	r1, r2, r3
2018-11-01 12:49:50,154: [26675] m.u.emulate:DEBUG: ==========
2018-11-01 12:49:50,154: [26675] m.u.emulate:DEBUG: Register R0   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,154: [26675] m.u.emulate:DEBUG: Register R1   Manticore: 00000000, Unicorn 000068ac
2018-11-01 12:49:50,154: [26675] m.u.emulate:DEBUG: Register R2   Manticore: 00001234, Unicorn 00001234
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R3   Manticore: 00005678, Unicorn 00005678
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R4   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R5   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R6   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R7   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R8   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R9   Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R10  Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R11  Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R12  Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R13  Manticore: 00010000, Unicorn 00010000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R14  Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register R15  Manticore: 00001000, Unicorn 00001000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: Register APSR  Manticore: 00000000, Unicorn 00000000
2018-11-01 12:49:50,155: [26675] m.u.emulate:DEBUG: >>>>>>>>>>
```

disconnect3d commented:
Thanks, that's a really good job here regarding arm things. Can you add it as a test?

NeatMonster commented:
@disconnect3d I've adapted my example script into a unit test.

disconnect3d commented on tests/test_unicorn.py:
```diff
@@ -1306,6 +1313,15 @@ def test_umull_z(self):
+        self.rf.write('R1', 0x1234)
+        self.rf.write('R2', 0x5678)
+        emulate_next(self.cpu)
+        self.assertEqual(self.rf.read('R0'), 0x1234 + 0x5678)
```
Hmm can we also add an assertion for cpu mode here?

NeatMonster replied:
I replaced your check on the Manticore cpu by a check on the Unicorn cpu mode. :smiley:
